_posts/2017-08-30-Modernizing-Force.markdown

```diff
@@ -0,0 +1,306 @@
+```
+
+```js
+// index.js
+
+const html = await renderLayout({
+  layout: 'templates/layout.handlebars',
+  data: {
+    title: 'Hello World!',
+  },
+  blocks: {
+    body: (props) => {
+      return (
+        <h1>
+          {props.title}
+        </h1>
+      )
+    }
+  }
+})
+
+console.log(html)
+
+// => Outputs:
+/*
+<html>
+  <head>
+    <title>Hello World!</title>
+  </head>
+  <body>
+    <h1>
+      Hello World!
+    </h1>
+  </body>
+</html>
+*/
+```
+
+In Force, we're using this to pattern to incrementally migrate portions of our app over to React, by taking existing block-based Jade layouts and injecting `ReactDOM.renderToString` output into them, and then rendering the layout into a HTML string that is passed down from the server and rehydrated on the client.
+
+Our existing Backbone views take advantage of the `templates` key:
+
+```js
+// server.js
+
+import LoginApp from 'components/LoginApp'
+
+const html = await renderLayout({
+  layout: 'templates/layout.handlebars',
+  data: {
+    title: 'Login / Sign-up',
+  },
+  blocks: {
+    app: LoginApp
+  },
+  templates: {
+    login: 'templates/login.jade'
+  }
+})
+
+res.send(html)
+```
+
+Similar to blocks, templates located in this object are pre-compiled and available to your components as `props.templates`.
+
+Once the html has been sent over the wire, we mount it like so:
+
+```js
+// client.js
+
+import LoginApp from 'components/LoginApp'
+
+React.render(
+  <LoginApp {...window.__BOOTSTRAP__} /> // Data passed down from `data` key
+)
+```
+
+```js
+// components/LoginApp.js
+
+import React from 'react'
+import LoginView from './LoginView'
+
+export default function LoginApp (props) {
+  const {
+    templates: {
+      login
+    }
+  } = props
+
+  return (
+    <LoginView
+      template={login}
+    />
+  )
+}
+```
+
+During the server-side render phase existing template code will be rendered with the component, and once the component is mounted on the client `componentDidMount` will fire and the Backbone view instantiated:
+
+```js
+// components/LoginView.js
+
+import React, { Component } from 'react'
+import LoginBackboneView from 'apps/login/views/LoginView'
+
+export default class LoginView extends Component {
+  componentDidMount () {
+    this.loginView = new LoginBackboneView()
+    this.loginView.render()
+  }
+
+  componentWillUnmount () {
+    this.loginView.remove()
+  }
+
+  render () {
+    return (
+      <div>
+        <div dangerouslySetInnerHtml={{
+          __html: this.props.template
+        }}>
+      </div>
+    )
+  }
+}
+```
+
+All of the possibilities that Stitch provides are too numerous to go over here, but check out the [documentation](https://github.com/artsy/stitch#usage) and [example apps](https://github.com/artsy/stitch/tree/master/examples) for more complete usage. While new, this pattern has worked quite well for us and has allowed Force to evolve alongside existing code with very little friction.
+
+[SOME KIND OF CLOSING PARAGRAPH]
```

Comment:
Loved it. I think your closing statements can aim to hit the same kind of notes as this sentence:

> But how? Given finite resources and a small team rebuilds can be difficult to execute, and so we had to figure out a way to marry the old with the new while also maintaining backwards compatibility / interoperability

Ideas: 
* Maybe do a CLOC on force to give people a sense of the size of codebase?
* There's definitely a common thread between this technique and the usage of React Native on the iOS side,  incremental revolutions over grand re-writes.
* The work to support future ideas improved working on older ideas also

* I wrote something [similar here](http://artsy.github.io/blog/2017/02/05/Retrospective-Swift-at-Artsy/):
 > Re-writing from scratch was not an option. That takes a lot of time and effort, which will happily remove technical debt, but that's not our issue. We also don't need or have a big redesign. However, a lot of companies used the Objective-C -> Swift transition as a time to re-write from scratch. We asked for the experiences from developers who had opted to do this, they said it was a great marketing tool for hiring - but was a lot of pain to actually work with day to day. They tend to talk about technical debt, and clean slates - but not that Objective-C was painful and Swift solves major architectural problems. With the notable exception of functional programming purists.